### PR TITLE
frame: support for serde added

### PR DESCRIPTION
**Pull request by michalkucharczyk**
- enabled `serde` features in dependent crates, no gate feature introduced, linker should do the job and strip unused code.
- frame::staking: added impl of `serde::Serialize, serde::Deserialize` for `enum Forcing`
- primitives::runtime: impl_opaque_keys macro provides `Serialize/Deserialize` impl if `serde` is enabled
- primitives::staking: added impl of `serde::Serialize`, `serde::Deserialize` for `enum StakerStatus`
- `serde::Serialize/Deserialize` implemented for pallets `GenesisConfig` in `no-std`

Step towards: paritytech/polkadot-sdk#25 

polkadot companion: paritytech/polkadot#7312
cumulus companion: paritytech/cumulus#2660

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2912216

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2912215

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2912217

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate-macos
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2916829

File: primitives/staking/Cargo.toml [hunk 1]
```diff
@@ -23,11 +23,14 @@ sp-std = { version = "6.0.0", default-features = false, path = "../std" }
...
+	"serde/std",
 	"codec/std",
 	"scale-info/std",
 	"sp-core/std",
 	"sp-runtime/std",
 	"sp-std/std",
 ]
+serde = [
+	"dep:serde"
+]
```
Comment (bkchr) [inline primitives/staking/Cargo.toml:35]:
```suggestion
```

As the crate is optional, we don't needan extra feature for it

**Review by bkchr — APPROVED** [review]

**Review by sam0x17 — APPROVED** [review]

**Review by davxy — APPROVED** [review]

**Review by acatangiu — APPROVED** [review]

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate-macos
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2927596

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate-macos
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2927651

**Review by melekes — APPROVED** [review]
👍

**Review by sam0x17 — APPROVED** [review]

**Comment by michalkucharczyk** [comment]
bot merge

**Comment by nazar-pc** [comment]
It is somewhat confusing that `serde` is an optional feature in some case, but essentially is not optional as far as runtime is concerned. We have custom `Block` struct implementation and for it `serde` has to be unconditionally enabled now because it is unconditionally enabled in `frame-support`, propagating `MaybeSerialize` requirement everywhere as the result, which was unfortunate as it slows down compilation.

**Comment by bkchr** [comment]
> which was unfortunate as it slows down compilation.

I mean this is expect-able, but how much does it slow down the compilation?

**Comment by nazar-pc** [comment]
Doing a/b testing would be tricky, hard to say the exact impact.

**Comment by xlc** [comment]
This is confusing to me. I don't really want all the extra code in my runtime to serialize/deserialize types. Those codes are never going to be used and bloat my wasm.

**Comment by koute** [comment]
> Those codes are never going to be used and bloat my wasm.

LTO should automatically remove most of the unused stuff. But assuming it somehow doesn't, can you show and measure how much exactly it bloats up your WASM blob?

**Comment by xlc** [comment]
I need to make my code compile before I can have numbers. And with https://github.com/paritytech/substrate/issues/13334 in mind, one of the goal is to be able to build genesis in wasm, so that those code won't be optimized out by LTO eventually.

But that brings up a good point, I guess we could have some CI task to measure the kitchen sink runtime wasm code size so we can see if a feature increased it significantly.

**Comment by michalkucharczyk** [comment]
> This is confusing to me. I don't really want all the extra code in my runtime to serialize/deserialize types. Those codes are never going to be used and bloat my wasm.

-> https://github.com/paritytech/substrate/issues/13334#issuecomment-1645187806

**Pull request merged**